tunnel/intra/retrier.go

```diff
@@ -0,0 +1,244 @@
+package intra
+
+import (
+	"io"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+// retrier implements the DuplexConn interface.
+type retrier struct {
+	// mutex is a lock that guards `conn`, `hello`, and `retryCompleteFlag`.
+	// These fields must not be modified except under this lock.
+	// After retryCompletedFlag is closed, these values will not be modified
+	// again so locking is no longer required for reads.
+	mutex   sync.Mutex
+	network string
+	addr    *net.TCPAddr
+	// conn is the current underlying connection.  It is only modified by the reader
+	// thread, so the reader functions may access it without acquiring a lock.
+	conn *net.TCPConn
+	// Time to wait between the first write and the first read before triggering a
+	// retry.
+	timeout time.Duration
+	// hello is the contents written before the first read.  It is initially empty,
+	// and is cleared when the first byte is received.
+	hello []byte
+	// Flag indicating when retry is finished or unnecessary.
+	retryCompleteFlag chan struct{}
+	// Flags indicating whether the caller has called CloseRead and CloseWrite.
+	readCloseFlag  chan struct{}
+	writeCloseFlag chan struct{}
+}
+
+// Helper functions for reading flags.
+// In this package, a "flag" is a thread-safe single-use status indicator that
+// starts in the "open" state and transitions to "closed" when close() is called.
+// It is implemented as a channel over which no data is ever sent.
+// Some advantages of this implementation:
+//  - The language enforces the one-way transition.
+//  - Nonblocking and blocking access are both straightforward.
+//  - Checking the status of a closed flag should be extremely fast (although currently
+//    it's not optimized: https://github.com/golang/go/issues/32529)
```

Comment:
I was trying to avoid using advanced synchronization features, since the "sync" package documentation recommends using channels instead when possible.

I also hope that this approach is less error-prone thanks to the one-way transition.